lib/src/rules/no_literal_bool_comparisons.dart

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+
+import '../analyzer.dart';
+
+const _desc = r"Don't compare booleans to boolean literals.";
+
+const _details = r'''
+From [Effective Dart](https://dart.dev/guides/language/effective-dart/usage#dont-use-true-or-false-in-equality-operations):
+
+**DON'T** use `true` or `false` in equality operations.
+
+**BAD:**
+```dart
+if (someBool == true) {
+}
+while (someBool == false) {
+}
+```
+
+**GOOD:**
+```dart
+if (someBool) {
+}
+while (!someBool) {
+}
+```
+''';
+
+class NoLiteralBoolComparisons extends LintRule {
+  NoLiteralBoolComparisons()
+      : super(
+          name: 'no_literal_bool_comparisons',
+          description: _desc,
+          details: _details,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    if (!context.isEnabled(Feature.non_nullable)) return;
+
+    var visitor = _Visitor(this, context);
+    registry.addBinaryExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+  final LinterContext context;
+
+  _Visitor(this.rule, this.context);
+
+  @override
+  void visitBinaryExpression(BinaryExpression node) {
```

Comment:
This also needs to check that the operator is either `==` or `!=`. This node is used for all binary operators, including `&&` and `||` (which arguably are just as bad, but I don't think this lint is intended to catch those cases).

There should also be at least one test to show that we don't report where we shouldn't.

Reply:
Oops, what a huge omission.
Done.

Reply:
It's easy to miss. `AssignmentExpression` has a similar issue; it's easy to forget about compound assignments.